### PR TITLE
4032 - UserInvitation and UserRole types and integration test

### PR DESCRIPTION
**Pull request by sorja**
Build issues etc can be ignored in this PR

Add unit test for getLastRole:

image

### CODE REVIEW

**Review by minotogna — CHANGES_REQUESTED** [review]
the comment in the previous PR should be updated.

File: src/meta/user/userRoles.ts [hunk 1]
```diff
@@ -50,10 +43,10 @@ const getLastRole = (params: { assessment?: Assessment; user: User }) => {
...
 
   const _roles = [...roles].sort((roleA, roleB) => {
-    if (!roleA.acceptedAt && !roleB.acceptedAt) return 0
-    if (!roleA.acceptedAt) return 1
-    if (!roleB.acceptedAt) return -1
-    return roleB.acceptedAt.localeCompare(roleA.acceptedAt)
+    if (!roleA.createdAt && !roleB.createdAt) return 0
+    if (!roleA.createdAt) return 1
+    if (!roleB.createdAt) return -1
+    return roleB.createdAt.localeCompare(roleA.createdAt)
```
Comment (minotogna) [inline src/meta/user/userRoles.ts]: should we use Dates instead ?

File: src/meta/user/userInvitations.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { UserInvitation } from 'meta/user/userInvitation'
+
+export const isInvitationExpired = (invitation: UserInvitation, expiryPeriod?: number) =>
+  new Date().getTime() - Date.parse(invitation.invitedAt) > (expiryPeriod || 7) * 86400000
+
+export const UserInvitations = {
+  isInvitationExpired,
```
Comment (minotogna) [inline src/meta/user/userInvitations.ts]: I would not duplicate the name invitation in the methods. this utility works on invitations objects only

File: src/meta/user/userInvitation.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { CountryIso } from 'meta/area'
+import { RoleName } from 'meta/user/userRole'
+
+export interface UserInvitation {
+  acceptedAt?: string
+  assessmentUuid?: string
+  countryIso?: CountryIso
+  cycleUuid: string
```
Comment (minotogna) [inline src/meta/user/userInvitation.ts]: how come cycle is required and assessment not ?

File: src/meta/user/userRole.ts [hunk 4]
```diff
@@ -53,18 +53,17 @@ export type UserRoleExtendedProps = UserRoleBaseProps & {
 }
 
 export interface UserRole<Name extends RoleName, Props extends UserRoleBaseProps = undefined, Permissions = undefined> {
-  id: number
   assessmentId?: number
-  cycleUuid: string
   countryIso?: CountryIso
-  invitedByUserUuid?: string
+  createdAt: string
+  cycleUuid: string
```
Comment (minotogna) [inline src/meta/user/userRole.ts]: same here... and are we still using assessment id ? or uuid ?

Comment (sorja) [reply src/meta/user/userRole.ts]: Typo, also cycle should be optional (admin)

File: src/meta/user/userInvitation.ts (hunk 3 above)
Comment (sorja) [reply src/meta/user/userInvitation.ts]: I will mark in the main PR that we should check if assessmentUuid is really needed as an invitation can be related only to a single cycle, same as roles

What do you think?

#discussion

Comment (minotogna) [reply src/meta/user/userInvitation.ts]: leave both, assessments and cycles for now. we'll talk in the near future about this. ok?

Comment (sorja) [reply src/meta/user/userInvitation.ts]: sure - added in main pr so we don't forget

also something to keep in mind in the discussion:

if role/invitation has either, it should maybe mean it's for all cycles (avoid role duplication?)

Comment (minotogna) [reply src/meta/user/userInvitation.ts]: an invitation is per cycle. same thing for the role. or did I misunderstood your comment ?

**Comment by minotogna** [comment]
please check deepscan , the tests have some issues

File: src/meta/user/userInvitation.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import { RoleName } from 'meta/user/userRole'
+
+export interface UserInvitation {
+  acceptedAt?: string
+  assessmentUuid: string
+  countryIso: CountryIso
+  cycleUuid: string
+  id: number
+  invitedAt: string
+  invitedByUserUuid?: string
```
Comment (minotogna) [inline src/meta/user/userInvitation.ts]: is it possible to have an invitation without the inviter ?

**Review by minotogna — APPROVED** [review]
👍

**Pull request merged**